File: .github/workflows/check_k8s_resources.js
const path = require('path/posix');
const fs = require('fs')
const execSync = require('child_process').execSync;
const YAML = require('yaml');


let resourced_kinds = ["Deployment", "StatefulSet", "Cronjob"];
let forbidden_kinds = ["Ingress", "SealedSecret"];
// `probelms` array is used to record failing criteria 
// to pvide a summary at the end of the ci
let problems = [];


// Maximum CPU and Memory to pass
let max_cpu = process.env.MAX_CPU || 10000;
let max_mem = process.env.MAX_MEM || 50000;

let cpu_usage = 0;
let mem_usage = 0;

// Assumed CPU and Memory limits if not defined
let default_cpu_limit = 1;
let default_mem_limit = "1Gi";


function main(chart_root) {
  console.log("Running with max cpu: " + max_cpu + " and max mem: " + max_mem);

  const chart_list = execSync(`ls ${path.join(chart_root)}`, { encoding: 'utf-8' });
  const active_apps = get_active_charts();
  let local_charts = chart_list.trim().split("\n");

  // Remove all charts that are not enabled in charts/apps
  let charts_to_template = local_charts.filter(chart => active_apps.includes(chart));
  // Workflows-cluster is manually added as it has no app as it is the cluster entry point
  charts_to_template.push("workflows-cluster")
  console.log("Templating the following charts: ")
  console.log(charts_to_template);

  charts_to_template.forEach(chart => check_chart(path.join(chart_root, chart)));
  console.log(`Total CPU usage: ${cpu_usage} mCPU`);
  console.log(`Total Memory usage: ${mem_usage} Mi`);

  verify_success();

  process.exit(0)
}

function verify_success() {
  // Explicitly checks against criteria 
  if (cpu_usage > max_cpu) {
    problems.push(`CPU usage exceeds the limit of ${max_cpu} mCPU`);
  }
  if (mem_usage > max_mem) {
    problems.push(`Memory usage exceeds the limit of ${max_mem} Mi`);
  }

  if (problems.length > 0) {
    console.log("Problems found:");
    console.log(problems);
    process.exit(1);
  }

  console.log("\nAll checks passed ✅");
  process.exit(0);
}

function get_active_charts() {
  const file = fs.readFileSync('./charts/apps/dev-values.yaml', 'utf8')
  let apps_yaml = YAML.parse(file);
  let enabled_charts = [];

  for (let key in apps_yaml) {
    if (apps_yaml[key].enabled) {
      enabled_charts.push(key);
    }
  }
  return enabled_charts;
}

function check_chart(chart_path) {
  // Purely cosmetic
  let header = `-----Checking chart ${chart_path}-----`;
  console.log("-".repeat(header.length));
  console.log(header);
  console.log("-".repeat(header.length));


  let full_chart = get_chart_template(chart_path);
  let sub_charts = full_chart.trim().split("---");

  sub_charts.forEach(sub_chart => {
    let parsed_chart = YAML.parse(sub_chart);
    if (parsed_chart == null) {
      return;
    }

    // Check for explicitly dis-allowed chart kinds eg ingresses
    if (forbidden_kinds.includes(parsed_chart.kind)) {
      console.log(`---`);
      console.log(`${parsed_chart.kind} found`);
      console.log(`Name: ${parsed_chart.metadata.name}`);
      problems.push(`${parsed_chart.kind} found in ${chart_path} with name ${parsed_chart.metadata.name}`);
      return;
    }

    // DFS through the chart yaml to find the resources.
    // Allows us to find resources without having to explicitly
    // know what kind of chart we are dealing with
    let resources = traverse_for_resources("resources", parsed_chart, "limits");
    let replicas = traverse_for_resources("replicas", parsed_chart, null);


    // If we find either a replicas or a resources,
    // we can assume this is a pod-creating chart.
    // If all criteria are not provided, we can fill in 
    // the missing components with the default cluster values
    let defined_resources = { all: true, cpu: true, mem: true };
    if (!resources && !replicas && resourced_kinds.includes(parsed_chart.kind)) {
      // This case is checked for a few kinds where we known they should hav 
      // resources etc but don't
      defined_resources.all = false;
    }
    if (!resources && !replicas) {
      return
    }

    if (!resources) {
      resources = { limits: { cpu: `${default_cpu_limit}`, memory: `${default_mem_limit}` } }
      defined_resources.all = false;
      defined_resources.cpu = false;
      defined_resources.mem = false;
    }
    if (!resources.limits.cpu) {
      resources.limits.cpu = `${default_cpu_limit}`;
      defined_resources.cpu = false;
    }
    if (!resources.limits.memory) {
      resources.limits.memory = `${default_mem_limit}`;
      defined_resources.mem = false;
    }
    if (!replicas) {
      replicas = 1
    }

    cpu_usage += fix_cpu_units(String(resources.limits.cpu)) * replicas;
    mem_usage += fix_mem_units(String(resources.limits.memory)) * replicas;

    console.log(`---`);
    console.log(`Name: ${parsed_chart.metadata.name}`);
    console.log(`Kind: ${parsed_chart.kind}`);
    console.log(`Resource Limits: CPU ${resources.limits.cpu}, Mem ${resources.limits.memory}`);
    console.log(`Replicas: ${replicas}`);

    if (!defined_resources.all) {
      console.log("No resources limits defined ⚠️  ");
    } else if (!defined_resources.cpu) {
      console.log("No resources limits for CPU defined ⚠️  ");
    } else if (!defined_resources.mem) {
      console.log("No resources limits for memory defined ⚠️  ");
    } else {
      console.log("All resource limits defined ✅");
    }
  });
  console.log("");
}

function fix_cpu_units(usage) {
  // Target unit in mCPU
  if (usage.endsWith("m")) {
    return Number(usage.slice(0, -1));
  }
  return Number(usage) * 1000;
}

function fix_mem_units(usage) {
  // Target unit Mi
  if (usage.endsWith("Mi")) {
    return Number(usage.slice(0, -2));
  }
  else if (usage.endsWith("Gi")) {
    return Number(usage.slice(0, -2)) * 1024;
  }
}

function traverse_for_resources(target_key, yaml, match_key) {
  if (yaml === null || typeof yaml !== 'object') {
    return undefined;
  }

  if (yaml.hasOwnProperty(target_key)) {
    // resource can appear in two different contexts so the match_key is used 
    // to check which is occuring
    if (match_key && yaml[target_key].hasOwnProperty(match_key) || !match_key) {
      return yaml[target_key];
    }
  }

  if (Array.isArray(yaml)) {
    for (let i = 0; i < yaml.length; i++) {
      const result = traverse_for_resources(target_key, yaml[i], match_key);
      if (result !== undefined) {
        return result;
      }
    }
  } else {

    for (let key in yaml) {
      if (key === "schema") {
        continue;
      }
      const result = traverse_for_resources(target_key, yaml[key], match_key);
      if (result !== undefined) {
        return result;
      }
    }
  }

  return undefined;
}

function get_chart_template(chart_path) {
  let command = `helm template test ${chart_path}`;

  if (fs.existsSync(`${chart_path}/dev-values.yaml`)) {
    command += ` -f ${chart_path}/dev-values.yaml`;
    console.log("Using dev values ✅");
  } else {
    console.log("No dev values could be found ❌");
  }

  return execSync(command, { encoding: 'utf-8', maxBuffer: 1000 * 1000 * 10 });
}

module.exports = main;
main("charts");

File: .github/workflows/check_k8s_resources.py
import yaml
import os
import subprocess
import textwrap
import re

ignored_kinds = ["ClusterPolicy", "Job"]
resourced_kinds = ["Deployment", "StatefulSet", "Cronjob"]
forbidden_kinds = ["Ingress", "SealedSecret"]
problems = []

max_cpu = int(os.getenv("MAX_CPU", 10000))
max_mem = int(os.getenv("MAX_MEM", 50000))

cumulative_cpu = 0
cumulative_mem = 0

default_cpu_limit = 1
default_mem_limit = "1Gi"


def main():
    print("Running with max cpu: " + str(max_cpu) + " and max mem: " + str(max_mem))
    all_local_charts = (
        subprocess.run("ls charts", capture_output=True, shell=True)
        .stdout.decode()
        .strip()
        .split("\n")
    )
    active_charts = get_active_charts()
    enabled_charts = list(filter(lambda x: x in active_charts, all_local_charts))
    enabled_charts.append("workflows-cluster")
    print(enabled_charts)

    [check_chart(chart) for chart in enabled_charts]

    header = "-----Report-----"
    print("-" * len(header))
    print(header)
    print("-" * len(header))
    if problems:
        print("Problems found:")
        print("\n".join(problems))
        exit(1)

    print("No problems found!")
    print("total cpu: " + str(cumulative_cpu) + " mCore")
    print("total mem: " + str(cumulative_mem) + " Mi")
    exit(0)


def check_chart(chart):
    chart_path = f"./charts/{chart}"
    header = f"-----Checking chart {chart_path}-----"
    print("-" * len(header))
    print(header)
    print("-" * len(header))

    full_chart = get_chart_yaml(chart_path)
    subcharts = [chart for chart in full_chart.strip().split("---") if len(chart) > 0]

    subchart_status = [check_subchart(chart_path, subchart) for subchart in subcharts]

    if cumulative_cpu > max_cpu:
        problems.append(f"CPU limit exceeded: {cumulative_cpu} > {max_cpu}")
    if cumulative_mem > max_mem:
        problems.append(f"Memory limit exceeded: {cumulative_mem} > {max_mem}")


def extract_resource_limits_and_replicas(manifest):
    results = []

    def traverse(sub_manifest):
        if isinstance(sub_manifest, dict):
            key = "containers"
            if key in sub_manifest and isinstance(sub_manifest[key], list):
                for container in sub_manifest[key]:
                    if isinstance(container, dict):
                        resources = container.get("resources", {})
                        limits = resources.get("limits", {})
                        results.append(limits)
            for value in sub_manifest.values():
                traverse(value)
        elif isinstance(sub_manifest, list):
            for item in sub_manifest:
                traverse(item)

    traverse(manifest)

    replicas = manifest.get("spec", {}).get("replicas", None)

    return {"replicas": replicas, "limits": results}


def check_subchart(parent_chart, raw_sub_chart):
    global cumulative_cpu
    global cumulative_mem
    curr_chart = get_subchart_yaml(raw_sub_chart)
    if curr_chart == 1:
        return 1

    if curr_chart["kind"] in forbidden_kinds:
        print("---")
        print(f"{curr_chart['kind']} found")
        print(f"Name: {curr_chart['metadata']['name']}")
        problems.append(
            f"{curr_chart['kind']} found in {parent_chart} with name {curr_chart['metadata']['name']}"
        )
        return
    if curr_chart["kind"] in ignored_kinds:
        return

    resources = extract_resource_limits_and_replicas(curr_chart)
    resources_defined = resources["limits"] or resources["replicas"]

    if not resources_defined and (curr_chart["kind"] in resourced_kinds):
        resources = {
            "replicas": 1,
            "limits": [{"cpu": default_cpu_limit, "memory": default_mem_limit}],
        }
    elif not resources_defined:
        return 0

    replicas = resources.get("replicas", 1)
    cpu = [container.get("cpu", default_cpu_limit) for container in resources["limits"]]
    memory = [
        container.get("memory", default_mem_limit) for container in resources["limits"]
    ]

    numeric_cpu = [format_resource("cpu", value) for value in cpu]
    numeric_memory = [format_resource("memory", value) for value in memory]
    pod_resources = list(zip(numeric_cpu, numeric_memory))

    cumulative_cpu += sum(numeric_cpu) * replicas
    cumulative_mem += sum(numeric_memory) * replicas

    print("---")
    print(f"Name: {curr_chart['metadata']['name']}")
    print(f"Kind: {curr_chart['kind']}")
    print(f"Replicas: {replicas}")
    for pod in pod_resources:
        print(f"  CPU: {pod[0]} mCore - Memory: {pod[1]} Mi")


def format_resource(resource_type, value):
    value = str(value)
    if resource_type == "cpu":
        if value.endswith("m"):
            return int(value[:-1])
        return int(value) * 1000

    if resource_type == "memory":
        if value.endswith("Mi"):
            return int(value[:-2])
        elif value.endswith("Gi"):
            return int(value[:-2]) * 1024
        print("value: " + value)
        raise Exception(f"Invalid memory unit\nValue:{value}")

    raise Exception(f"Invalid resource type\n{value}")


def get_active_charts():
    with open("./charts/apps/dev-values.yaml") as f:
        values = yaml.safe_load(f)
        active_charts = [key for key in values if values[key]["enabled"]]
        return active_charts


def get_chart_yaml(path):
    command = f"helm template workflows-cluster {path}"
    if os.path.exists(f"{path}/dev-values.yaml"):
        command += f" -f {path}/dev-values.yaml"
        print("Using dev values ✅")
    else:
        print("No dev values could be found ❌")

    response = subprocess.run(command, capture_output=True, shell=True)
    if response.returncode != 0:
        raise Exception(f"Error running helm template\n{response.stderr.decode()}")

    return response.stdout.decode()


def get_subchart_yaml(raw_yaml):
    try:
        dedented = textwrap.dedent(raw_yaml)
        cleaned_placeholders = re.sub(r"{{.*?}}", "placeholder", dedented)
        parsed = yaml.safe_load(cleaned_placeholders)
        return parsed
    except Exception as err:
        print(f"Error parsing yaml {err}")
        print(raw_yaml)
        return 1


main()
